packages/babel-core/src/config/files/plugins.ts

```diff
@@ -92,63 +93,116 @@ function standardizeName(type: "plugin" | "preset", name: string) {
+    }
+  };
+}
+function tryCatchAsync<Fn extends (...args: any) => Promise<any>>(
+  fn: Fn,
+): (...args: Parameters<Fn>) => Promise<Result<Awaited<ReturnType<Fn>>>> {
+  return async (...args) => {
+    try {
+      // @ts-ignore
+      return { error: null, value: await fn(...args) };
+    } catch (error) {
+      return { error, value: null };
+    }
+  };
+}
+
+function* resolveAlternativesHelper(
   type: "plugin" | "preset",
   name: string,
-  dirname: string = process.cwd(),
-) {
+): Iterator<string, string, Result<string>> {
   const standardizedName = standardizeName(type, name);
+  const { error, value } = yield standardizedName;
+  if (!error) return value;
 
-  try {
-    return require.resolve(standardizedName, {
-      paths: [dirname],
-    });
-  } catch (e) {
-    if (e.code !== "MODULE_NOT_FOUND") throw e;
-
-    if (standardizedName !== name) {
-      let resolvedOriginal = false;
-      try {
-        require.resolve(name, {
-          paths: [dirname],
-        });
-        resolvedOriginal = true;
-      } catch {}
-
-      if (resolvedOriginal) {
-        e.message += `\n- If you want to resolve "${name}", use "module:${name}"`;
-      }
-    }
+  // @ts-ignore
+  if (error.code !== "MODULE_NOT_FOUND") throw error;
 
-    let resolvedBabel = false;
-    try {
-      require.resolve(standardizeName(type, "@babel/" + name), {
-        paths: [dirname],
-      });
-      resolvedBabel = true;
-    } catch {}
-
-    if (resolvedBabel) {
-      e.message += `\n- Did you mean "@babel/${name}"?`;
-    }
+  if (standardizedName !== name && !(yield name).error) {
+    error.message += `\n- If you want to resolve "${name}", use "module:${name}"`;
+  }
 
-    let resolvedOppositeType = false;
-    const oppositeType = type === "preset" ? "plugin" : "preset";
-    try {
-      require.resolve(standardizeName(oppositeType, name), {
-        paths: [dirname],
-      });
-      resolvedOppositeType = true;
-    } catch {}
-
-    if (resolvedOppositeType) {
-      e.message += `\n- Did you accidentally pass a ${oppositeType} as a ${type}?`;
-    }
+  if (!(yield standardizeName(type, "@babel/" + name)).error) {
+    error.message += `\n- Did you mean "@babel/${name}"?`;
+  }
 
-    throw e;
+  const oppositeType = type === "preset" ? "plugin" : "preset";
+  if (!(yield standardizeName(oppositeType, name)).error) {
+    error.message += `\n- Did you accidentally pass a ${oppositeType} as a ${type}?`;
   }
+
+  throw error;
 }
 
+const tryRequireResolve = tryCatchSync(require.resolve);
+const tryImportMetaResolve = tryCatchAsync(importMetaResolve);
+
+function resolveStandardizedNameForRequrie(
+  type: "plugin" | "preset",
+  name: string,
+  dirname: string,
+) {
+  const it = resolveAlternativesHelper(type, name);
+  let res = it.next();
+  while (!res.done) {
+    res = it.next(tryRequireResolve(res.value, { paths: [dirname] }));
+  }
+  return res.value;
+}
+async function resolveStandardizedNameForImport(
```

Comment:
I kept these two functions separated (rather than creating a gensync) because they are not simply a sync and an async version: the sync one is also used when Babel is called asynchronously.